scipy/signal/filter_design.py

```diff
@@ -328,22 +344,57 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
     >>> plt.show()
 
     """
-    b, a = map(atleast_1d, (b, a))
-    if whole:
-        lastpoint = 2 * pi
-    else:
-        lastpoint = pi
+    b = atleast_1d(b)
+    a = atleast_1d(a)
+
     if worN is None:
-        N = 512
-        w = numpy.linspace(0, lastpoint, N, endpoint=False)
-    elif isinstance(worN, int):
-        N = worN
-        w = numpy.linspace(0, lastpoint, N, endpoint=False)
+        worN = 512
+
+    h = None
+    try:
+        worN = operator.index(worN)
+    except TypeError:  # not int-like
+        w = atleast_1d(worN)
     else:
-        w = worN
-    w = atleast_1d(w)
-    zm1 = exp(-1j * w)
-    h = polyval(b[::-1], zm1) / polyval(a[::-1], zm1)
+        lastpoint = 2 * pi if whole else pi
+        w = np.linspace(0, lastpoint, worN, endpoint=False)
+        min_size = max(a.shape[axis], b.shape[axis])
+        if worN >= min_size and fftpack.next_fast_len(worN) == worN:
+            # if worN is fast, 2 * worN will be fast, too,
+            # so no need to check that one
+            n_fft = worN if whole else worN * 2
+            if np.isrealobj(b) and np.isrealobj(a):
+                h = np.fft.rfft(b, n=n_fft, axis=axis)
+                # don't do inplace division so we can use broadcasting
+                if a.shape[axis] == 1:  # shortcut speedup
+                    h = h / a
+                else:
+                    h = h / np.fft.rfft(a, n=n_fft, axis=axis)
+                trunc = [slice(None)] * h.ndim
+                trunc[axis] = slice(worN)
+                h = h[trunc]
+                if whole:
+                    # exclude DC and maybe Nyquist (no need to use axis_reverse
+                    # here because we can build reversal with the truncation)
+                    stop = h.shape[axis] - 1 if n_fft % 2 == 1 else -2
+                    h_flip = [slice(None)] * h.ndim
+                    h_flip[axis] = slice(stop, 0, -1)
+                    h = np.concatenate((h, h[h_flip].conj()), axis=axis)
+            else:  # we can use complex FFT
+                h = fftpack.fft(b, n=n_fft, axis=axis)[:worN]
+                if a.size == 1:
+                    h /= a
+                else:
+                    h /= fftpack.fft(a, n=n_fft, axis=axis)[:worN]
+    del worN
+
+    if h is None:  # still need to compute using freqs w
+        zm1 = exp(-1j * w)
+        bb = np.apply_along_axis(polyval, axis,
```

Comment:
`npp_polyval` is the same as `polyval` in the imports at the top, if that's what you mean.

If you're asking why didn't I use the `swapaxes` snippet you posted, I tried it at first but it didn't work for the multidimensional cases (broadcasting errors), whereas `np.apply_along_axis` did.